streams/jvm/src/main/scala/zio/stream/experimental/platform.scala

```diff
@@ -119,4 +125,288 @@ trait ZStreamPlatformSpecificConstructors {
+  final def fromJavaStream[R, A](stream: => java.util.stream.Stream[A]): ZStream[R, Throwable, A] =
+    ZStream.fromJavaIterator(stream.iterator())
+
+  /**
+   * Creates a stream from a Java stream
+   */
+  final def fromJavaStreamEffect[R, A](
+    stream: ZIO[R, Throwable, java.util.stream.Stream[A]]
+  ): ZStream[R, Throwable, A] =
+    ZStream.fromJavaIteratorEffect(stream.flatMap(s => UIO(s.iterator())))
+
+  /**
+   * Creates a stream from a managed Java stream
+   */
+  final def fromJavaStreamManaged[R, A](
+    stream: ZManaged[R, Throwable, java.util.stream.Stream[A]]
+  ): ZStream[R, Throwable, A] =
+    ZStream.fromJavaIteratorManaged(stream.mapM(s => UIO(s.iterator())))
+
+  /**
+   * Creates a stream from a Java stream
+   */
+  final def fromJavaStreamTotal[A](stream: => java.util.stream.Stream[A]): ZStream[Any, Nothing, A] =
+    ZStream.fromJavaIteratorTotal(stream.iterator())
+
+  /**
+   * Create a stream of accepted connection from server socket
+   * Emit socket `Connection` from which you can read / write and ensure it is closed after it is used
+   */
+  def fromSocketServer(
+    port: Int,
+    host: Option[String] = None
+  ): ZStream[Any, Throwable, Connection] =
+    for {
+      server <- ZStream.managed(ZManaged.fromAutoCloseable(ZIO.effectBlocking {
+                  AsynchronousServerSocketChannel
+                    .open()
+                    .bind(
+                      host.fold(new InetSocketAddress(port))(new InetSocketAddress(_, port))
+                    )
+                }))
+
+      registerConnection <- ZStream.managed(ZManaged.scope)
+
+      conn <- ZStream.repeatEffect {
+                ZIO
+                  .effectAsync[Any, Throwable, UManaged[Connection]] { callback =>
+                    server.accept(
+                      null,
+                      new CompletionHandler[AsynchronousSocketChannel, Void]() {
+                        self =>
+                        override def completed(socket: AsynchronousSocketChannel, attachment: Void): Unit =
+                          callback(ZIO.succeed(Connection.make(socket)))
+
+                        override def failed(exc: Throwable, attachment: Void): Unit = callback(ZIO.fail(exc))
+                      }
+                    )
+                  }
+                  .flatMap(managedConn => registerConnection(managedConn).map(_._2))
+              }
+    } yield conn
+
+  /**
+   * Accepted connection made to a specific channel `AsynchronousServerSocketChannel`
+   */
+  class Connection(socket: AsynchronousSocketChannel) {
+
+    /**
+     * Read the entire `AsynchronousSocketChannel` by emitting a `Chunk[Byte]`
+     * The caller of this function is NOT responsible for closing the `AsynchronousSocketChannel`.
+     */
+    def read: ZStream[Any, Throwable, Byte] =
+      ZStream.unfoldChunkM(0) {
+        case -1 => ZIO.succeed(Option.empty)
+        case _ =>
+          val buff = ByteBuffer.allocate(ZStream.DefaultChunkSize)
```

Comment:
We could use a single `ByteBuffer` repeatedly as in the other stream readers instead of allocating one for each `socket.read`